### PR TITLE
Add support for armv7l.

### PR DESCRIPTION
**Pull request by jsirois**

### CODE REVIEW

**Comment by jsirois** [comment]
@Lauszus thanks for getting the ball rolling in https://github.com/a-scie/jump/pull/256.
I think this approach using `cross` is probably a bit cleaner, but I'd like your feedback. In particular, unlike in your change, I'm just rolling with `env::consts::ARCH` here for armv7l which means the binary suffix for arm 32 is just `arm`. I'd like to make sure we agree on a suffix, because I'll want ptex, scie-jump and science to all use the same one.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -15,42 +15,65 @@ jobs:
...
+            name: Linux x86-64
+          - os: ubuntu-24.04
+            name: Linux x86-64 (musl)
+            cross-target: x86_64-unknown-linux-musl
+          - os: ubuntu-24.04
+            name: Linux aarch64
+            cross-target: aarch64-unknown-linux-musl
+          - os: ubuntu-24.04
+            name: Linux armv7l
+            cross-target: armv7-unknown-linux-musleabihf
```
Comment (jsirois) [inline .github/workflows/ci.yml:41]: @Lauszus note that no cargo config is needed for linker flags, just specifying the musl targets is enough to get you a statically linked executable:
```
:; file dist/ptex-linux-* | grep -v sha256
dist/ptex-linux-aarch64:        ELF 64-bit LSB executable, ARM aarch64, version 1 (SYSV), statically linked, stripped
dist/ptex-linux-arm:            ELF 32-bit LSB executable, ARM, EABI5 version 1 (SYSV), statically linked, stripped
dist/ptex-linux-x86_64:         ELF 64-bit LSB pie executable, x86-64, version 1 (SYSV), static-pie linked, stripped
```

I built those on my GNU-Linux x86-64 box using `cross run -p package --target {x86_64-unknown-linux-musl,aarch64-unknown-linux-musl,armv7-unknown-linux-musleabihf}`.

**Review by benjyw — APPROVED** [review]
Looks great, about as streamlined as I can imagine it being. And I concur that `cross-rs` is likely the way to go (and it looks like they officially bless tonistiigi/binfmt...)

**Comment by Lauszus** [comment]
> @Lauszus thanks for getting the ball rolling in [a-scie/jump#256](https://github.com/a-scie/jump/pull/256). I think this approach using `cross` is probably a bit cleaner, but I'd like your feedback. In particular, unlike in your change, I'm just rolling with `env::consts::ARCH` here for armv7l which means the binary suffix for arm 32 is just `arm`. I'd like to make sure we agree on a suffix, because I'll want ptex, scie-jump and science to all use the same one.

Yeah that is better. I will update my PR to use `cross` as well.

I don't think we should use `arm` for the executable, as it is very ambiguous. Anything else I have ever seen either uses `armv7l` or `aarch64`. This is also what Python standalone uses and what is used for Raspberry Pi wheels.

File: README.md [hunk 2]
```diff
@@ -141,10 +141,10 @@ distribution is ~28M:
...
 
-The `ptex` binary is [released](https://github.com/a-scie/ptex/releases) for Linux (x86_64 &
-aarch64), macOS (x86_64 & aarch64) and Windows (x86_64). If you'd like to build your own copy,
-you'll need [Rust installed](https://rustup.rs/) at which point you can run `cargo run -p package`
-and a binary for your current machine will be built in `dist/` along with a sha256 checksum file.
-For more build options, you can run `cargo run -p package -- --help`. On some systems, builds will
-require `cmake`, `make` and `perl` in order to build various `*-sys` crates. If you're missing
-these, the build failures will point you in the right direction with some reading.
+The `ptex` binary is [released](https://github.com/a-scie/ptex/releases) for Linux (x86_64,
+aarch64 & armv7), macOS (x86_64 & aarch64) and Windows (x86_64 & aarch64). If you'd like to build
```
Comment (Lauszus) [inline README.md:145]:
```suggestion
aarch64 & armv7l), macOS (x86_64 & aarch64) and Windows (x86_64 & aarch64). If you'd like to build
```

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -15,42 +15,65 @@ jobs:
...
       - name: Check Formatting
+        if: matrix.cross-target == ''
         run: |
           rustup toolchain add nightly -c rustfmt
           cargo +nightly fmt --check --all
+      - name: Install Cross
+        if: matrix.cross-target != ''
+        run: |
+          cargo install cross --locked
+          docker run --privileged --rm tonistiigi/binfmt --install all
```
Comment (Lauszus) [inline .github/workflows/ci.yml:61]: Is this step really needed? You don't do this in the release workflow?

File: CHANGES.md [hunk 4]
```diff
@@ -1,5 +1,9 @@
 # Release Notes
 
+## 1.3.0
+
+This release adds support for Linux ARM (armv7 and armv8 32 bit mode).
```
Comment (Lauszus) [inline CHANGES.md:5]:
```suggestion
This release adds support for Linux ARM (armv7l and armv8l 32 bit mode).
```

File: .github/workflows/ci.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (jsirois) [reply .github/workflows/ci.yml:61]: Yes. Although the cross images contain a cross-compiler toolchain for each target, that's only 1/2 the battle for commands like `cargo test` and `cargo run`, the other 1/2 is actually running the binary that is compiled and that requires qemu / binfmt support.

**Comment by jsirois** [comment]
> I don't think we should use arm for the executable, as it is very ambiguous. Anything else I have ever seen either uses armv7l or aarch64. This is also what Python standalone uses and what is used for Raspberry Pi wheels.

Ok, fixed up to use `armv7l` like you did in the scie jump.

If this goes green I'll merge and test out the release process. If that goes well, I'll swing back to your similar scie jump change and get that released and then work on science and then Pex.

Thanks again @Lauszus.

**Comment by Lauszus** [comment]
> > I don't think we should use arm for the executable, as it is very ambiguous. Anything else I have ever seen either uses armv7l or aarch64. This is also what Python standalone uses and what is used for Raspberry Pi wheels.
> 
> Ok, fixed up to use `armv7l` like you did in the scie jump.
> 
> If this goes green I'll merge and test out the release process. If that goes well, I'll swing back to your similar scie jump change and get that released and then work on science and then Pex.
> 
> Thanks again @Lauszus.

You are welcome. I can send an initial PR for pex and science if you want, as I started on it, but then stopped as I was waiting for a new jump and ptex release.

**Comment by jsirois** [comment]
> I can send an initial PR for pex and science if you want, as I started on it, but then stopped as I was waiting for a new jump and ptex release.

That sounds great.

**Comment by Lauszus** [comment]
> > I can send an initial PR for pex and science if you want, as I started on it, but then stopped as I was waiting for a new jump and ptex release.
> 
> That sounds great.

I opened two draft PRs: https://github.com/a-scie/lift/pull/106 and https://github.com/pex-tool/pex/pull/2620. I'll just clean them up a bit.

**Pull request merged**